File: kernel/src/interrupts/pic.c
#include "pic.h"
#include "../io.h"

#define PIC1    0x20
#define PIC2    0xA0

#define PIC1_COMMAND	PIC1
#define PIC1_DATA	(PIC1+1)
#define PIC2_COMMAND	PIC2
#define PIC2_DATA	(PIC2+1)

void IRQ_set_mask(uint8_t IRQline) {
    uint16_t port;
    uint8_t value;

    if(IRQline < 8) {
        port = PIC1_DATA;
    } else {
        port = PIC2_DATA;
        IRQline -= 8;
    }
    value = inb(port) | (1 << IRQline);
    outb(port, value);        
}

void IRQ_clear_mask(uint8_t IRQline) {
    uint16_t port;
    uint8_t value;

    if(IRQline < 8) {
        port = PIC1_DATA;
    } else {
        port = PIC2_DATA;
        IRQline -= 8;
    }
    value = inb(port) & ~(1 << IRQline);
    outb(port, value);        
}

#define ICW1_ICW4	0x01		/* Indicates that ICW4 will be present */
#define ICW1_SINGLE	0x02		/* Single (cascade) mode */
#define ICW1_INTERVAL4	0x04		/* Call address interval 4 (8) */
#define ICW1_LEVEL	0x08		/* Level triggered (edge) mode */
#define ICW1_INIT	0x10		/* Initialization - required! */

#define ICW4_8086	0x01		/* 8086/88 (MCS-80/85) mode */
#define ICW4_AUTO	0x02		/* Auto (normal) EOI */
#define ICW4_BUF_SLAVE	0x08		/* Buffered mode/slave */
#define ICW4_BUF_MASTER	0x0C		/* Buffered mode/master */
#define ICW4_SFNM	0x10		/* Special fully nested (not) */

/*
arguments:
	offset1 - vector offset for master PIC
		vectors on the master become offset1..offset1+7
	offset2 - same for slave PIC: offset2..offset2+7
*/
void PIC_remap(int offset1, int offset2)
{
	outb(PIC1_COMMAND, ICW1_INIT | ICW1_ICW4);  // starts the initialization sequence (in cascade mode)
	io_wait();
	outb(PIC2_COMMAND, ICW1_INIT | ICW1_ICW4);
	io_wait();
	outb(PIC1_DATA, offset1);                 // ICW2: Master PIC vector offset
	io_wait();
	outb(PIC2_DATA, offset2);                 // ICW2: Slave PIC vector offset
	io_wait();
	outb(PIC1_DATA, 4);                       // ICW3: tell Master PIC that there is a slave PIC at IRQ2 (0000 0100)
	io_wait();
	outb(PIC2_DATA, 2);                       // ICW3: tell Slave PIC its cascade identity (0000 0010)
	io_wait();
	
	outb(PIC1_DATA, ICW4_8086);               // ICW4: have the PICs use 8086 mode (and not 8080 mode)
	io_wait();
	outb(PIC2_DATA, ICW4_8086);
	io_wait();

	// Mask both PICs.
	for(int i = 0; i < 16; i++){
        IRQ_set_mask(i);
        io_wait();
    }
}

#define PIC_EOI		0x20

void pic_sendEOI(uint8_t irq)
{
    if (irq >= 0x28)				// if the IRQ came from the slave PIC
        outb(PIC2_COMMAND, PIC_EOI);

    outb(PIC1_COMMAND, PIC_EOI);		// if the IRQ came from the master and/or the slave PIC
}

File: kernel/rust-kernel/src/interrupts/pic.rs
const PIC1: u16 = 0x20;
const PIC2: u16 = 0xA0;

const PIC1_COMMAND: u16 = PIC1;
const PIC1_DATA: u16 = PIC1 + 1;
const PIC2_COMMAND: u16 = PIC2;
const PIC2_DATA: u16 = PIC2 + 1;

use crate::{inb, io_wait, outb};

#[unsafe(no_mangle)]
pub fn IRQ_set_mask(mut irq_num:  u8){
    let port = if irq_num < 8 {
        PIC1_DATA
    }else {
        irq_num -= 8;
        PIC2_DATA
    };
    unsafe{
        let value = inb(port) | (1 << irq_num);
        outb(port, value);
    }
}

#[unsafe(no_mangle)]
pub fn IRQ_clear_mask(mut irq_num:  u8){
    let port = if irq_num < 8 {
        PIC1_DATA
    }else {
        irq_num -= 8;
        PIC2_DATA
    };
    unsafe{
        let value = inb(port) & !(1 << irq_num);
        outb(port, value);
    }
}

const ICW1_ICW4: u8 = 0x01;
const ICW1_INIT: u8 = 0x10;

#[unsafe(no_mangle)]
pub fn PIC_remap(offset1: u8, offset2: u8){
    unsafe{
        outb(PIC1_COMMAND, ICW1_INIT | ICW1_INIT); //start the initialization sequence
        io_wait();
        outb(PIC1_COMMAND, ICW1_INIT | ICW1_INIT);
        io_wait();

        // set the offsets
        outb(PIC1_DATA, offset1);
        io_wait();
        outb(PIC2_DATA, offset2);
        io_wait();

        outb(PIC1_DATA, 4);
        io_wait();
        outb(PIC2_DATA, 2);
        io_wait();

        // Mask both PICs
        for i in 0..16{
            IRQ_set_mask(i);
            io_wait();
        } 
    }
}

const PIC_EOI: u8 = 0x20;

#[unsafe(no_mangle)]
pub fn PIC_sendEOI(irq: u8){
    unsafe {
        if irq >= 0x28 {
            outb(PIC2_COMMAND, PIC_EOI); // interrupts from the slave
        }
        // we need to send end of interrupt to the master in both case
        outb(PIC1_COMMAND, PIC_EOI);
    }
}